numba/types/function.py

```diff
@@ -0,0 +1,116 @@
+
+__all__ = ['FunctionType', 'FunctionProtoType', 'WrapperAddressProtocol']
+
+from .abstract import Type, Literal
+
+
+class FunctionType(Type):
+    """The front-end for first-class function types.
+    """
+
+    mutable = True
+    cconv = None
+
+    def __init__(self, ftype):
+        if isinstance(ftype, tuple) and len(ftype) == 2:
+            ftype = FunctionProtoType(ftype[0], ftype[1])
+        if isinstance(ftype, FunctionProtoType):
+            self.ftype = ftype
+            name = ftype.name
+        else:
+            raise TypeError(
+                f'constructing {type(self).__name__} from {type(ftype)}'
+                ' instance is not supported')
+        super(FunctionType, self).__init__(name)
+
+    def signature(self):
+        from numba import typing
+        ptype = self.ftype
+        return typing.signature(ptype.rtype, *ptype.atypes)
+
+    @property
+    def key(self):
+        return self.name
+
+    def cast_python_value(self, value):
+        from numba import typing
+        if isinstance(value, typing.Signature):
+            ptype = FunctionProtoType(value.return_type, value.args)
+            return FunctionType(ptype)
+        raise NotImplementedError(
+            'cast_python_value({}, {})'.format(value, type(value)))
+
+    def get_call_type(self, context, args, kws):
+        from numba import typing
+        ptype = self.ftype
+        if len(args) == len(ptype.atypes):
+            for i, atype in enumerate(args):
+                if isinstance(atype, Literal):
+                    atype = atype.literal_type
+                if not (atype <= ptype.atypes[i]):
```

Comment:
Not all `Type` instances are ordered. 

Below is a failing case because `Array` is not ordered:

<details>
<summary>example code</summary>

```python
# Failed because Array type-class doesn't have <= comparison
import numpy as np
from numba import njit, cfunc, intp, float64, function

@cfunc("intp(intp[:], float64[:])")
def foo1(x, y):
    return x[0] + y[0]


@cfunc("intp(intp[:], float64[:])")
def foo2(x, y):
    return x[0] - y[0]

@njit
def bar(fx, fy, i):
    a = np.array([10], dtype=np.intp)
    b = np.array([12], dtype=np.float64)
    if i == 0:
        f = fx
    elif i == 1:
        f = fy
    else:
        return
    return f(a, b)

r = bar(foo1, foo2, 0)
print(r)
```
</details>

Reply:
Fixed. The currently supported argument types are arrays, numbers, and literals.